Advanced Search Options

avocadodev123 opened the pull request:
### **Advanced Search Function for Library & Database Cleanup Cron Job**

_**Description:**_

This PR introduces an Advanced Search Function for the library, allowing users to filter publications more effectively. The new search functionality includes:

1. Search by: author, tags, ISBN, title, abstract.

2. Combination search: Supports AND, OR, NOT for refined results.

3. Search modes: Contains, Contains exact phrase, and Starts with to enhance precision.

4. Improved UI/UX: Added CSS transition effects for a smoother and more polished user experience.

5. Search visualization: Provides a better overview of search results.

Additionally, this PR implements a cron job that runs every 6 hours to clean up duplicate entries in the Zotero and MOI Web Library databases. This ensures data integrity in case the duplicate prevention mechanism fails during the syncing process.

**_How to Test:_**
1. Navigate to /islands/library/.

2. Click on the Advanced Search option.

3. Test various search filters and combinations to verify the functionality.

4. Ensure smooth UI transitions and visualization elements display correctly.

5. (Optional) Verify the database cleanup by checking for duplicate entries before and after the cron job runs.

KentTran2302 committed:
working on add more fields in advanced option search through library

KentTran2302 committed:
working on add fields for advanced search

KentTran2302 committed:
working on handle the queries in the backend for library items

KentTran2302 committed:
improve advance-search

KentTran2302 committed:
working on advanced search options for library

KentTran2302 committed:
added view for advanced search in library

KentTran2302 committed:
edit gitignore

KentTran2302 committed:
working on adding fields

KentTran2302 committed:
fix bugs for front-end advanced search

KentTran2302 committed:
working on crons.py

KentTran2302 committed:
remove the test for the Zotero Item

paulhoekman reviewed (changes requested):
Thanks Kent but this one needs better testing it seems. I tried it out searching for publications published before 2000 and it still showed me publications from e.g. 2023. 

![Screenshot_2025-03-27_08-12-07](https://github.com/user-attachments/assets/2b301bb4-86d6-4a3f-bed5-defc008eca7a)

Please do full testing of the feature and let me know when it's ready for review.

There is another feature also included here which is the feature to clean the library of duplicates. Is the problem that there are duplicate items at the moment? If so, I suggest that you create this code inside the control panel and you allow the user to clean up the library like that. Or you just run it once-off through a view. But running it as a cron means we have a structural problem with the duplicate-checking code and we should address that instead of constantly cleaning it up - do you agree? Please add that code as a separate PR to discuss separately, but please include it differently from a constantly running cron. Thanks!

avocadodev123 commented:
> Thanks Kent but this one needs better testing it seems. I tried it out searching for publications published before 2000 and it still showed me publications from e.g. 2023.
> 
> ![Screenshot_2025-03-27_08-12-07](https://private-user-images.githubusercontent.com/4469695/427369417-2b301bb4-86d6-4a3f-bed5-defc008eca7a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxMjYyNTYsIm5iZiI6MTc0MzEyNTk1NiwicGF0aCI6Ii80NDY5Njk1LzQyNzM2OTQxNy0yYjMwMWJiNC04NmQ2LTRhM2YtYmVkNS1kZWZjMDA4ZWNhN2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjhUMDEzOTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmRmNTZkODg1OWU3MTE0N2QxZDgyOWYxYTMzMmE0YTYwZjVmYTBlYTYyYWFlNWEyMmVjNmIzZDZlOWU1OWE0YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.aDAgMaAIRx2FMX7TQ-5hX8uJxtW4XpA6p83PmFE90-w)
> 
> Please do full testing of the feature and let me know when it's ready for review.
> 
> There is another feature also included here which is the feature to clean the library of duplicates. Is the problem that there are duplicate items at the moment? If so, I suggest that you create this code inside the control panel and you allow the user to clean up the library like that. Or you just run it once-off through a view. But running it as a cron means we have a structural problem with the duplicate-checking code and we should address that instead of constantly cleaning it up - do you agree? Please add that code as a separate PR to discuss separately, but please include it differently from a constantly running cron. Thanks!

Thanks Paul, I will look into this issue.

KentTran2302 committed:
fix search without advanced filters

avocadodev123 commented:
Hi @paulhoekman, I have investigated the problem. This happens because if no terms are being input below, it will assume that the search is just a simple search and I did not add the filtering by dates on the simple search. I have fixed that issue and I have tested it. I also tested with Albert and Simron on the advanced search filtering with some testcases, so if you tried that out and got incorrect result, please let me know.
 
I also took the clean up library from the crons job and I will put it in another PR so we can discuss more about this.

paulhoekman commented on src/core/static/css/main.css:
```diff
@@ -1459,9 +1459,140 @@ body.library.search .select2-container--bootstrap4 .select2-dropdown {
   border: none; }
 
+/* Styling for the expand-advanced-search-bar */
+body.library.search .expand-advanced-search-bar {
```
Please confirm: is the CSS only applicable to a single HTML page, or to multiple? If it's a single page only, please include directly in that particular page. If it's multiple, it's fine in the main.css file.

paulhoekman commented on src/library/views.py:
```diff
@@ -54,34 +242,96 @@ def index(request):
+        booleans = request.GET.getlist('boolean')
+
+        print("Fields got: ", fields)
+        print("Contains got: ", contains)
```
Please remove all the print() statements - I assume that was for debugging purposes?

paulhoekman commented on src/library/views.py:
```diff
@@ -54,34 +242,96 @@ def index(request):
+
         if not request.GET.get("urban_only"):
-            urban_only = False
+                urban_only = False
```
Why is the indentation changed here?

paulhoekman commented on src/templates/_library.list.html:
```diff
@@ -34,7 +34,7 @@
               {{ each }}
             </a>
-          </td>
+          </td>        
```
Please undo this extra space change - doesn't seem intentional...?

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -266,15 +712,557 @@ <h2>Download or export</h2>
+    theme: "bootstrap4",
+  });
+
+  // Event listener for the advanced search button click
```
It's a bit of an odd mix between jquery syntax and non-jquery syntax. Pls use jquery consistently, especially instead of using `document.querySelector` in so many places. Thanks.

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -266,15 +712,557 @@ <h2>Download or export</h2>
+}
+
+  // simple search button to redirect the user back to the simple search page
+  document.getElementById("simple-search-btn").addEventListener("click", function() {
```
Another example (but there are many, please review fully). Write this like so please:

$("#simple-search-btn").on("click", function() {

paulhoekman reviewed (changes requested):
Thanks Kent. Again no time for testing right now but I just did a quick read through of the code and added some comments. Once all addressed I will look at functionality. Thanks.

KentTran2302 committed:
fix PR and using jQuery consistently

avocadodev123 commented:
Thanks Paul for the insights, I have fixed the code based on your comments and changed some major javascript function to jQuery consistently. Please let me know if anything arises, if something still not in jQuery, just ping me again and let me know that there is still something need to be converted. Other than that, thank you very much!!

paulhoekman commented:
Hi Kent, I'm just closing the PRs in the old repo. I have merged this code into the new repo when I set that up. New repo: https://github.com/Metabolism-of-Islands/metabolism-of-islands-platform
